### PR TITLE
Fixed #27996 - Added crypto extension and function to contrib.postgres

### PR DESCRIPTION
**Pull request by pauloxnet**
Added CryptoExtension (pgcrypto) extension and RandomUUID
(GEN_RANDOM_UUID) function to contrib.postgres

Ticket [#27996](https://code.djangoproject.com/ticket/27996)

### CODE REVIEW

File: docs/ref/contrib/postgres/functions.txt [hunk 1]
```diff
@@ -28,3 +28,17 @@ Usage example::
...
     <QuerySet [<Article: How to Django>]>
+
+``RandomUUID``
+==================
+
+.. class:: RandomUUID()
+
+.. versionadded:: 2.0
+
+Returns a version 4 (random) UUID.
```
Comment (mjtamlyn) [inline docs/ref/contrib/postgres/functions.txt]: This needs to state that it needs the extension installing.

Also wouldn't hurt to include the canonical example code.

Even better, let's test whether `UUIDField(default=RandomUUID)` works - if it does let's document that both here and in the UUID docs as a note for PG users. (sounds like it doesn't but let's verify)

File: docs/releases/2.0.txt [hunk 2]
```diff
@@ -76,6 +76,11 @@ Minor features
   :class:`~django.contrib.postgres.aggregates.ArrayAgg` determines if
   concatenated values will be distinct.
 
+* Added the :class:`~django.contrib.postgres.functions.RandomUUID` database
```
Comment (MarkusH) [inline docs/releases/2.0.txt]: I think this deserves 2 bullet points. One for `RandomUUID` and one for `pgcrypto`.

File: tests/postgres_tests/test_functions.py [hunk 3]
```diff
@@ -26,3 +27,20 @@ def test_transaction_now(self):
...
+
+class TestRandomUUID(PostgreSQLTestCase):
+
+    def test_random_uuid(self):
+        m1 = UUIDTestModel.objects.create()
+        m2 = UUIDTestModel.objects.create()
+
+        UUIDTestModel.objects.filter(id=m1.id).update(uuid=RandomUUID())
+        sleep(0.1)
+        UUIDTestModel.objects.filter(id=m2.id).update(uuid=RandomUUID())
```
Comment (mjtamlyn) [inline tests/postgres_tests/test_functions.py]: This test would be better if we updated the whole model with the expression, like in the PR

File: tests/postgres_tests/migrations/0001_setup_extensions.py [hunk 4]
```diff
@@ -22,9 +23,10 @@ class Migration(migrations.Migration):
         BtreeGinExtension(),
         # Ensure CreateExtension quotes extension names by creating one with a
         # dash in its name.
+        CITextExtension(),
```
Comment (MarkusH) [inline tests/postgres_tests/migrations/0001_setup_extensions.py]: Can you please leave this line at the end or make sure that the comment above still goes with the `CreateExtension('uuid-ossp')` line.

File: tests/postgres_tests/test_functions.py [hunk 5]
```diff
@@ -26,3 +27,20 @@ def test_transaction_now(self):
...
+
+
+class TestRandomUUID(PostgreSQLTestCase):
+
+    def test_random_uuid(self):
+        m1 = UUIDTestModel.objects.create()
+        m2 = UUIDTestModel.objects.create()
+
+        UUIDTestModel.objects.filter(id=m1.id).update(uuid=RandomUUID())
+        sleep(0.1)
```
Comment (MarkusH) [inline tests/postgres_tests/test_functions.py]: Why do you need this sleep function? If `RandomUUID` returns the same UUID for two consecutive calls at the same second that seems to be a bug in PostgreSQL.

File: docs/ref/contrib/postgres/functions.txt (hunk 1 above)
Comment (MarkusH) [inline docs/ref/contrib/postgres/functions.txt]: I think you can omit `(random)` here.

File: tests/postgres_tests/test_functions.py (hunk 5 above)
Comment (pauloxnet) [inline tests/postgres_tests/test_functions.py]: Actually it's only a dead code I forgot to delete, I'm going to remove it.

File: tests/postgres_tests/migrations/0001_setup_extensions.py (hunk 4 above)
Comment (pauloxnet) [inline tests/postgres_tests/migrations/0001_setup_extensions.py]: You're right I alphabetical ordered the operation but I forgot to move the comment too. I'm fixing it.

File: docs/releases/2.0.txt [hunk 6]
```diff
@@ -76,6 +76,13 @@ Minor features
   :class:`~django.contrib.postgres.aggregates.ArrayAgg` determines if
   concatenated values will be distinct.
 
+* The new :class:`~django.contrib.postgres.functions.RandomUUID` database
+  function returns a version 4 UUID.
+
+* The new :class:`~django.contrib.postgres.operations.CryptoExtension`
+  migration operation allow using PostgreSQL's ``pgcrypto`` extension that
```
Comment (Ian-Foote) [inline docs/releases/2.0.txt]: ~~allow~~ allows

Comment (pauloxnet) [inline docs/releases/2.0.txt]: Thanks, I'm going to fix it.

File: django/contrib/postgres/functions.py [hunk 7]
```diff
@@ -1,4 +1,13 @@
-from django.db.models import DateTimeField, Func
+from django.db.models import DateTimeField, Func, UUIDField
+
+
+class RandomUUID(Func):
+    template = 'GEN_RANDOM_UUID()'
+
+    def __init__(self, output_field=None, **extra):
+        if output_field is None:
+            output_field = UUIDField()
```
Comment (orf) [inline django/contrib/postgres/functions.py]: I think you could use `_output_field = UUIDField()` at the class level instead of this?

Comment (MarkusH) [inline django/contrib/postgres/functions.py]: That's slightly too much magic than I'd like, I think. Thoughts on that @jarshwah?

File: docs/ref/contrib/postgres/functions.txt [hunk 8]
```diff
@@ -28,3 +28,25 @@ Usage example::
...
+==================
+
+.. class:: RandomUUID()
+
+.. versionadded:: 2.0
+
+The ``RandomUUID`` function returns a version 4 UUID using a dedicated
+PostgreSQL extension.
+
+To use it, add ``'django.contrib.postgres'`` in your :setting:`INSTALLED_APPS`
```
Comment (MarkusH) [inline docs/ref/contrib/postgres/functions.txt]: I believe referencing `INSTALLED_APPS` here is not necessary.

File: django/contrib/postgres/functions.py (hunk 7 above)
Comment (pauloxnet) [inline django/contrib/postgres/functions.py]: I used the same convention that I found in the same file for `TransactionNow` function.

File: docs/ref/contrib/postgres/functions.txt (hunk 8 above)
Comment (pauloxnet) [inline docs/ref/contrib/postgres/functions.txt]: I tried to align this note to the same note I found into `lookups.txt`

File: django/contrib/postgres/functions.py (hunk 7 above)
Comment (MarkusH) [inline django/contrib/postgres/functions.py]: Then it's settled, I think. Thanks :)

File: docs/ref/contrib/postgres/functions.txt (hunk 8 above)
Comment (MarkusH) [inline docs/ref/contrib/postgres/functions.txt]: Hmm ...  Could you stick to the style of the current file?

Comment (MarkusH) [inline docs/ref/contrib/postgres/functions.txt]: "To use `RandomUUID` you have to have the pgcrypto extension installed."

Comment (MarkusH) [inline docs/ref/contrib/postgres/functions.txt]: But I'm happy to defer to @timgraham for suggestions.

Comment (pauloxnet) [inline docs/ref/contrib/postgres/functions.txt]: I'm going to modify it as you suggested.

**Comment by pauloxnet** [comment]
I'm waiting for the tests to be run on all databases.

**Comment by pauloxnet** [comment]
All checks have passed, I hope this will merged.

**Pull request merged**